README
INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

Documentation is provided in the module and perldoc output is copied below:

NAME
       Redis::hiredis − interact with Redis using the hiredis client.

SYNOPSIS
         use Redis::hiredis;
         my $redis = Redis::hiredis−>new();
         $redis−>connect('127.0.0.1', 6379);
         $redis−>command('set foo bar');
         $redis−>command(["set", "foo", "bar baz"]); # values with spaces
         my $val = $redis−>command('get foo');

         # to pipeline commands
         $redis−>append_command('set abc 123');
         $redis−>append_command('get abc');
         my $set_status = $redis−>get_reply(); # 'OK'
         my $get_val = $redis−>get_reply(); # 123

DESCRIPTION
       "Redis::hiredis" is a simple wrapper around Salvatore Sanfilippo’s
       hiredis C client that allows connecting and sending any command just
       like you would from a command line Redis client.

       NOTE Versions >= 0.9.2 and <= 0.9.2.4 are not compatible with prior
       versions

       METHODS


       new([utf8 => 1], [host => "localhost"], [port => 6379])
           Creates a new Redis::hiredis object.

           If the host attribute is provided the "connect" method will
           automatically be called.

       connect( $hostname, $port )
           $hostname is the hostname of the Redis server to connect to

           $port is the port to connect on.  Default 6379

       command( $command_and_args )
       command( [ $command, $arg, ... ] )
       command( $command, $arg, ... )
           command supports multiple types of calls to be backwards compatible
           and provide more convenient use.  Examples of how to pass arguments
           are:

             $redis−>command('set foo bar');
             $redis−>command(["set", "foo", "bar baz"]);
             $redis−>command("set", "foo", "bar baz");

           Note that if you have spaces in your values, you must use one of
           the last 2 forms.

           command will return a scalar value which will either be an integer,
           string or an array ref (if multiple values are returned).

       append_command( $command )
           For performance reasons, it’s sometimes useful to pipeline
           commands.  When pipelining, muiltple commands are sent to the
           server at once and the results are read as they become available.
           hiredis supports this via append_command() and get_reply().
           Commands passed to append_command() are buffered locally until the
           first call to get_reply() when all the commands are sent to the
           server at once.  The results are then returned one at a time via
           calls to get_reply().

           See the hiredis documentation for a more detailed explanation.

       get_reply()
           See append_command().

       Autoloaded Methods

       Autoload is used to allow an interface like $redis−>set("foo", "bar").
       The method name you provide will be passed blindly to Redis, so any
       supported command should work.

       Note that to use any autoloaded method, you must pass arguments as an
       array, the string and array ref forms supported by command will not
       work.

SEE ALSO
       The Redis command reference can be found here:
       <http://redis.io/commands>

       A discusion of pipelining can be found here:
       <http://redis.io/topics/pipelining>

       Documentation on the hiredis client can be found here:
       <https://github.com/antirez/hiredis>

       Redis::hiredis on github: <https://github.com/neophenix/redis−hiredis>


Changes
0.14.1  - Sat Apr 11 02:58:00 EDT 2020
        * Bump hiredis version to 14.1

0.11.0  - Tue Apr 02 10:09:00 EDT 2013
        * Cache AUTOLOAD functions by Songmu
        * Bump hiredis version to latest by Songmu

0.10.2  - Tue Jun 12 11:06:00 EDT 2012
        * update net.c to latest bits from hiredis, mostly to pull in change to use poll and not select

0.10.1  - Sat Apr 07 10:46:00 EDT 2012
        * update hiredis lib to 10.1
        * connect_unix additions by Toru Yamaguchi
        * file relocation / renaming

0.9.2.8 - Sat Jul 23 11:17:00 EDT 2011
        * don't croak on REDIS_REPLY_ERROR if we are processing a multi bulk reply since they can have failures and successes intertwined

0.9.2.7 - Thu Jun 16 16:59:00 EDT 2011
        * croak on a null reply, patch via Eugene Loktionov

0.9.2.6 - Mon Feb 28 20:43:00 EST 2011
        * update doc links

0.9.2.5 - Sat Feb 19 21:02:00 EST 2011
        * command supports array of args now too command("set", "foo", "bar")
        * Added autoload to allow calls like set("foo", "bar") and get("foo")

0.9.2.4 - Sat Feb 19 11:59:00 EST 2011
        * command supports an array ref of args again, fixes issue where values, etc could not contain spaces
        * utf8 and "connect on new" support from David Leadbeater

0.9.2.3 - Sat Jan 08 20:16:00 EST 2011
        * Remove Test::Exception from tests, no need to make it a dependency if I don't have to

0.9.2.2 - Mon Jan 03 09:47:00 EST 2011
        * Patch from Whitney Jackson to provide pipelining, better error handling and code modularization

0.9.2.1 - Fri Dec 24 10:11:00 EST 2010
        * Fix DESTROY to not free NULL pointer

0.9.2 - Thu Dec 23 15:51:00 EST 2010
        * Update to latest hiredis release
        * Bump version to now match hiredis

0.0.3 - Tue Nov 02 11:01:00 EDT 2010
        * Bring lib-hiredis.c up to date
        * Fix bug #62234, seg fault on return of array of integers
        * Added tests for bug #62234

0.0.2 - Tue Aug 31 17:05:00 EDT 2010
        * Fix tests to not fallback to localhost and skip if
            REDISHOST isn't defined

0.0.1 - Sun Aug 29 10:48:00 EDT 2010
        * Initial release


lib/Redis/hiredis.pm
package Redis::hiredis;

use strict;
our $VERSION = "0.14.1";
require XSLoader;
XSLoader::load('Redis::hiredis', $VERSION);

our $AUTOLOAD;

sub new {
    my($class, %args) = @_;

    if(!exists $args{utf8}) {
        $args{utf8} = 1;
    }

    my $self = $class->_new($args{utf8});

    if(exists $args{host}) {
        $self->connect($args{host}, defined $args{port} ? $args{port} : 6379);
    }
    elsif (exists $args{path}) {
        $self->connect_unix($args{path});
    }

    return $self;
}

sub AUTOLOAD {
    (my $method = $AUTOLOAD) =~ s/.*:://;

    # cache method for future calls
    my $sub =  sub { shift->command($method, @_) };
    no strict 'refs';
    *$AUTOLOAD = $sub;

    goto $sub;
}

1;
__END__

=head1 NAME

Redis::hiredis - interact with Redis using the hiredis client.

=head1 SYNOPSIS

  use Redis::hiredis;
  my $redis = Redis::hiredis->new();
  $redis->connect('127.0.0.1', 6379);
  $redis->command('set foo bar');
  $redis->command(["set", "foo", "bar baz"]); # values with spaces
  my $val = $redis->command('get foo');

  # to pipeline commands
  $redis->append_command('set abc 123');
  $redis->append_command('get abc');
  my $set_status = $redis->get_reply(); # 'OK'
  my $get_val = $redis->get_reply(); # 123

=head1 DESCRIPTION

C<Redis::hiredis> is a simple wrapper around Salvatore Sanfilippo's
L<hiredis|https://github.com/antirez/hiredis> C client that allows connecting 
and sending any command just like you would from a command line Redis client.

B<NOTE> Versions >= 0.9.2 and <= 0.9.2.4 are not compatible with prior versions

=head2 METHODS

=over 4

=item new([utf8 => 1], [host => "localhost"], [port => 6379], [path => "/tmp/redis.sock"])

Creates a new Redis::hiredis object.

If the host attribute is provided the L</connect> method will automatically be
called.

If the path attribute is provided the L</connect_unix> method will automatically be
called.

=item connect( $hostname, $port )

C<$hostname> is the hostname of the Redis server to connect to

C<$port> is the port to connect on.  Default 6379

=item connect_unix( $path )

C<$path> is the path to the unix socket

=item command( $command_and_args )

=item command( [ $command, $arg, ... ] )

=item command( $command, $arg, ... )

command supports multiple types of calls to be backwards compatible and provide
more convenient use.  Examples of how to pass arguments are:

  $redis->command('set foo bar');
  $redis->command(["set", "foo", "bar baz"]);
  $redis->command("set", "foo", "bar baz");

Note that if you have spaces in your values, you must use one of the last 2 forms.

command will return a scalar value which will either be an integer, string
or an array ref (if multiple values are returned).

=item append_command( $command )

For performance reasons, it's sometimes useful to pipeline commands.  When
pipelining, muiltple commands are sent to the server at once and the results
are read as they become available.  hiredis supports this via append_command()
and get_reply().  Commands passed to append_command() are buffered locally
until the first call to get_reply() when all the commands are sent to the
server at once.  The results are then returned one at a time via calls to
get_reply().

See the hiredis documentation for a more detailed explanation.

=item get_reply()

See append_command().

=back

=head2 Autoloaded Methods

Autoload is used to allow an interface like $redis->set("foo", "bar").  The method
name you provide will be passed blindly to Redis, so any supported command should work.

Note that to use any autoloaded method, you must pass arguments as an array, the string
and array ref forms supported by command will not work.

=head1 SEE ALSO

The Redis command reference can be found here:
L<http://redis.io/commands>

A discusion of pipelining can be found here:
L<http://redis.io/topics/pipelining>

Documentation on the hiredis client can be found here:
L<https://github.com/antirez/hiredis>

Redis::hiredis on github:
L<https://github.com/neophenix/redis-hiredis>

=cut
